wolfcrypt/src/evp.c

```diff
@@ -6878,6 +6878,416 @@ int wolfSSL_EVP_get_hashinfo(const WOLFSSL_EVP_MD* evp,
+    outsz = BASE64_ENCODE_RESULT_BLOCK_SIZE + 1; /* 64 byte and one for LF*/
+    res = Base64_Encode(ctx->data, ctx->remaining ,out, &outsz);
+    if (res == 0)
+        *outl = outsz;
+    else
+        *outl = 0;
+
+    ctx->remaining = 0;
+    XMEMSET(ctx->data, 0, sizeof(ctx->data));
+
+    return;
+}
+#endif /* WOLFSSL_BASE64_ENCODE */
+#if defined(WOLFSSL_BASE64_DECODE)
+
+/*  wolfSSL_EVP_DecodeInit initializes specified WOLFSSL_EVP_ENCODE_CTX struct
+ *  for subsequent wolfSSL_EVP_DecodeUpdate.
+ */
+void wolfSSL_EVP_DecodeInit(WOLFSSL_EVP_ENCODE_CTX* ctx)
+{
+    WOLFSSL_ENTER("wolfSSL_EVP_DecodeInit");
+    /* clean up ctx */
+    if (ctx != NULL) {
+        ctx->remaining = 0;
+        XMEMSET(ctx->data, 0, sizeof(ctx->data));
+    }
+}
+/*  wolfSSL_EVP_DecodeUpdate encodes the input data in 4-byte units
+ *  and outputs it to out. If less than 4 bytes of data remain, save it in
+ *  ctx. The data given in the subsequent wolfSSL_EVP_DecodeUpdate
+ *  is combined with the data stored in CTX and used for decoding.
+ *  Returns 1 or 0 on success, -1 on error. Return value 0 indicates that
+ *  clients should call wolfSSL_EVP_DecodeFinal as next call.
+ */
+int  wolfSSL_EVP_DecodeUpdate(WOLFSSL_EVP_ENCODE_CTX* ctx,
+            unsigned char* out, int* outl, const unsigned char* in, int inl)
+{
+    word32 outsz = 0;
+    word32 j = 0;
+    word32 inLen;
+    int    res;
+    int    pad = 0;
+    int    i;
+    int cpySz;
+    unsigned char c;
+    int pad3 = 0;
+    int pad4 = 0;
+    byte e[4];
+
+    WOLFSSL_ENTER("wolfSSL_EVP_DecodeUpdate");
+
+    if (outl == NULL)
+        return -1;
+
+    if (ctx == NULL || out == NULL || in == NULL) {
+        *outl = 0;
+        return -1;
+    }
+
+    if (inl == 0) {
+        *outl = 0;
+        return 1;
+    }
+
+    inLen = inl;
+    *outl = 0;
+
+    /* if the remaining data exist in the ctx, add input data to them to create
+    a block(4bytes) for decoding*/
+    if ( ctx->remaining > 0 && inl > 0) {
+
+        cpySz = min((BASE64_DECODE_BLOCK_SIZE - ctx->remaining), inl);
+
+        for ( i = 0; cpySz > 0 && inLen > 0; i++) {
+            if ((res = Base64_SkipNewline(in, &inLen, &j))
+                                            == ASN_INPUT_E) {
+                return -1;  /* detected an illegal char in input */
+            }
+            c = in[j++];
+
+            if (c == '=')
+                pad = 1;
+
+            *(ctx->data + ctx->remaining + i) = c;
+            inLen--;
+            cpySz--;
+        }
+
+        outsz = sizeof(ctx->data);
+        res = Base64_Decode( ctx->data, BASE64_DECODE_BLOCK_SIZE, out, &outsz);
+        if (res == 0) {
+            *outl += outsz;
+            out   += outsz;
+
+            ctx->remaining = 0;
+            XMEMSET(ctx->data, 0, sizeof(ctx->data));
+        }
+        else {
+            *outl = 0;
+            return -1;   /* return with error */
+        }
+    }
+
+    /* process data in input buffer */
+    while (inLen > 3) {
```

Comment:
This looks very similar to what Base64_Decode() is doing internally.  Are you sure we need to also do this here before calling Base64_Decode()?

Reply:
That is needed and the reason is as follows:
Base64_Decode is not a streaming process, so it processes the input data and exits. If a line break or whitespace character is found in the input data, it will be skipped, but if the end point of the input data is reached as a result, Base64_Decode will stop processing there. The data cleansing is required before Base64_Decode so that the processing does not stop within 4 bytes, which is the unit of Base64 decoding processing. The logic that exists before calling Base64_Decode in a while-loop is a data cleansing process that removes line breaks and whitespaces.